RavenDB-6716 - Ready to merge - fast test is passing.

fitzchak opened the pull request:
Need to fix all failing tests because that we now version also deleted documents.

Commit:
RavenDB-6716 Better naming, change ExpirationActive to HasExpirationConfiguration

Commit:
RavenDB-6716 Put a revision also for document's delete.
Put a delete marker on the last delete revision so we can query them. We'll need to clear it on put.
Delete duplicate GetRevisionCommand.

Commit:
RavenDB-6716 If we already using StringBuilder we should use it all the way

Commit:
RavenDB-7371 Use the overload method

Commit:
RavenDB-6716 Fix failing tests

Commit:
RavenDB-6716 Remove WIP class that should be removed

Commit:
RavenDB-6716 Change delete-documents-with-revisions to revisions/zombied

Commit:
RavenDB-6716 Change how we get the zombied revision in order to optimize Put method over the Get method. The way to get zombied revisions is: /databases/*/revisions/zombied?etag={long.MaxValue}&pageSize=25

Commit:
RavenDB-6716 GetLongQueryString should return long instead of long? when it is mandatory field.

ayende commented on src/Raven.Server/Documents/Handlers/VersioningHandler.cs:
```diff
@@ -143,5 +140,45 @@ private Task GetRevisions()
+            {
+                var etag = GetLongQueryString("etag");
+                var pageSize = GetPageSize(); // TODO: limit page size. By default it is int.MaxValue!
+                var revisions = versioningStorage.GetZombiedRevisions(context, etag, pageSize).ToArray();
```
Why are you forcing us to materialize it all at one shot? There is no need to do that.
ETag check can be done in the same way we do it for the `/docs` endpoint.

fitzchak replied:
I'll implement this.

fitzchak replied:
This will be done in a different PR

ayende commented on src/Raven.Server/Documents/Handlers/VersioningHandler.cs:
```diff
@@ -143,5 +140,45 @@ private Task GetRevisions()
+                    return Task.CompletedTask;
+                }
+
+                HttpContext.Response.Headers["ETag"] = actualEtag.ToString();
```
Don't we need to use `"` there?

fitzchak replied:
Fixed.

Commit:
RavenDB-6716 Fix test and make EnsureMetadata method support ZombiedRevision

Commit:
RavenDB-6716 Fix merge

Commit:
RavenDB-6716 Keep the old change vector of the imported document. This way we won't have two revisions of the same document, as both of them will have the same change vector.

Commit:
RavenDB-6716 Fix a serious bug where IncrementCountOfRevisions would not be called and revisions count would be out of sync.
Also make sure that we honor MaxRevisions also when deleting a document and inserting a DeleteRevision.

Commit:
RavenDB-6716 Change DocumentFlags.ZombiedRevision to DocumentFlags.DeleteRevision as it will be shown for the user in the document also for not zombied revisions.

Commit:
RavenDB-6716 Fix test. Replicate the collection of a deleteRevision. Use the correct change vector for the deleteRevision, so it will be the same after replication.

Commit:
RavenDB-6716 Do not increment the revisions count if we just update the same revision

Commit:
RavenDB-6716 Fix test. oldDoc can be null when coming from smuggler import.

Commit:
RavenDB-6716 Fix tests.

fitzchak commented:
@ayende Ready to merge

Commit:
RavenDB-6716 Avoid calling ToArray to get the latest etag

ayende commented on src/Raven.Server/Documents/Document.cs:
```diff
@@ -38,15 +41,26 @@ public unsafe ulong DataHash
         }
 
-        public void EnsureMetadata(float? indexScore = null)
+        private static readonly DynamicJsonValue ZombiedRevision = new DynamicJsonValue();
```
This is not safe for multi threaded use, why not just create it as needed?

fitzchak replied:
Fixed

ayende commented on src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs:
```diff
@@ -832,8 +832,11 @@ public override int Execute(DocumentsOperationContext context)
+
+                                    // no need to load document data for tombstones
+                                    // document size == -1 --> doc is a tombstone
+                                    // document size == 0 --> doc is deleteRevision
```
Better to reserve 0 and user -1, -2, etc.

fitzchak replied:
Done
